### ISSUE TITLE
Querying with custom types

### ISSUE THREAD

**Issue by adamburgess** [mikro-orm/mikro-orm#739]
**Is your feature request related to a problem? Please describe.**
Using custom types currently works when inserting/updating, but I can't _query_ with them.

I want to use this for Date <--> unix timestamp (number).
**Describe the solution you'd like**
```ts
@Entity()
class Foo {
    @PrimaryKey()
    id!: number

    @Property({ type: UnixTimestamp })
    bar!: Date
}

class UnixTimestamp extends Type<Date, number> {
    convertToDatabaseValue(value: Date | number) {
        console.log('converting', value);
        if (value instanceof Date)
            return value.getTime();
        return value;
    }
    convertToJSValue(value: Date | number) {
        if (value instanceof Date)
            return value;
        return new Date(value);
    }
    getColumnType() {
        return 'number';
    }
}

em.find(Foo, { bar: new Date(1000) });
// converting 1970-01-01T00:00:01.000Z
// [query] select `e0`.* from `foo` as `e0` where `e0`.`bar` = 1000

```

Currently:
```ts
// [query] select `e0`.* from `foo` as `e0` where `e0`.`bar` = '1970-01-01 00:00:01.000'
```
**Describe alternatives you've considered**
It does work doing something like:
```ts
em.find(Foo, { bar: new UnixTimestamp().convertToDatabaseValue(new Date(1000)) as unknown as Date });
```
... which provides the correct type... but I'd prefer not to do this...

**Additional context**
This example here is not necessary for databases other than sqlite as they have working datetimes, but say I instead used a truly custom JS type such as dayjs or moment instance, then I'd always have to do the conversion manually

**Comment by B4nan** [comment]
I do understand what you want, but doing `new Date(1000)` is definitely wrong.

**Comment by adamburgess** [comment]
In what way? I'm simply constructing a date object. I'd hope to be able to use any object, provided there's a transform to/from database types.

**Comment by B4nan** [comment]
Well technically it is ok, but timestamp 1000 (ms) is really 1970-01-01 00:00:01.000, that is the part that looks wrong to me.

**Comment by adamburgess** [comment]
Ah ok, yeah I pulled that 1 second past 1970 out of a hat to demonstrate. Maybe I should've chosen one closer to the current date? ;)

**Comment by B4nan** [comment]
Btw I was playing with dates and sqlite recently, and based on my tests, `Date` objects are correctly converted to timestamps at the sqlite3 level, so you might see the query with string/iso date, but in fact it should work like that just fine.

Just tested:

```
// this will result in query with iso date printed
const t1 = await orm.em.getRepository(Author4).findOne({ createdAt: { $gte: new Date('1990-03-24') } });

// the query:
select `e0`.* from `author4` as `e0` where `e0`.`created_at` >= '1990-03-24 01:00:00.000' limit 1
```

But in the database they are timestamps (and the query returns the results correctly):

![image](https://user-images.githubusercontent.com/615580/91152795-b3ecbd00-e6bf-11ea-9774-6f8ad986e43f.png)

So basically there is no need for a `UnixTimestamp` type, it's actually the opposite, if you want to store datetimes in sqlite as strings, you will be facing this issue #761.

**Comment by adamburgess** [comment]
Sqlite is pretty magic, then.

I tested now and it seems you're completely correct:
```ts
@Property({ columnType: 'integer' })
date!: Date
```
is all that's needed for sqlite!

So, my issue is solved, so I guess this can be closed?
Unless you'd rather me come up with another example – that doesn't use dates – and make another issue.

**Comment by B4nan** [comment]
For the sqlite/date this is not needed, but I'd say it's a valid issue in general. Imagine you would use a `Point` class that would be mapped to string `[x,y]` or something. `FilterQuery` will require using the `Point` class in your conditions, but the final query would most probably fail (or it might depend on the `toJSON` implementation probably).

So let's keep this open (or feel free to open new issue with different example), I will take a closer look at this `Point` example to see how it actually behaves.
